docs/investigations/simplify-api-usage/simplify-api-usage.md

```diff
@@ -0,0 +1,141 @@
+# Simplify usage of Compass API
+
+## Terminology:
+`consumer` - client of a Compass API (Application, Runtime, Integration System)
+
+## Overview
+
+This document discusses how to simplify usage of Compass API
+
+As of right now, there are some mutations that are related to a base entity (e.g. Application or Runtime) and they require the ID of the base entity:
+ 
+`addAPI(applicationID: ID!, in: APIDefinitionInput! @validate): APIDefinition! @hasScopes(path: "graphql.mutation.addAPI")` 
+
+needs `applicationID`, and
+
+`setRuntimeLabel(runtimeID: ID!, key: String!, value: Any!): Label! @hasScopes(path: "graphql.mutation.setRuntimeLabel")`
+ 
+ needs `runtimeID`.
+
+So when an application wants to add API for itself, it needs to provide its own ID. That behaviour is inconvenient and should be changed.
+Although the option to provide an consumerType ID should stay. Take for example a case when an Integration System creates an Application.
+
+There is another case when Integration System creates an application, it also has to provide its own ID to the `integrationSystemID` parameter inside `ApplicationCreateInput`.
+
+
+## Requirements
+* `consumer id (e.g. applicationID)` not required
+
+## Possible solutions
+
+### 1. `consumer ID` should be optional for operations related to itself
+ 
+The `applicationID`/`runtimeID` fields could be changed to be optional in mutations in the schema
+ 
+The idea is to create a directive which will fetch the ID of the consumer when it's not present.
+It is possible due to the fact that it can be retrieved from the request context.
+The directive could be used both on parameter(e.g. `addAPI` mutation) and field inside input(`integrationSystemID` case).
+Also, it would be possible for a consumer to update itself (`updateApplication` mutation) and not provide it's `ID`
+ 
+**Work that has to be done**
+* prepare directive which will retrieve `ID` depending on consumer type
+* change `ID` parameters to optional in mutations in GraphQL schema
+* adjust resolvers
+
+**Pros**
+* no existing code needs to change
+
+**Cons**
+* A LOT work with tests, see that example: 
+
+considering `updateApplication` mutation, we would have to write tests to cover these cases:
```

Comment:
currently, our security concept that defines who can execute which mutations is based only on scopes. In the near future, we have to implement more restrictive checks, that application cannot update data of the other Application. So then, a similar list of tests has to be implemented. But implementing viewer pattern right now provide some value: we can simplify life of Application provider - they don't have to store ID, but they can query it every time.